File: Dart/1-50/12.dart
import 'dart:io';

void main() {
  var n = int.parse(stdin.readLineSync()!);
  var a = List<int>.generate(n + 1, (i) => i);

  a[1] = 0;
  var i = 2;

  while (i <= n) {
    if (a[i] != 0) {
      var j = i + i;
      while (j <= n) {
        a[j] = 0;
        j += i;
      }
    }
    i++;
  }

  var primes = a.where((num) => num != 0).toList();
  var ism = ["Bobur", "Ali"];

  if (n == 1) {
    print("Ali");
  } else {
    print(ism[primes.length % 2]);
  }
}

File: Golang/1-50/12.go
package main

import (
	"fmt"
)

func main() {
	var n int
	fmt.Scan(&n)

	a := make([]int, n+1)
	for i := 0; i <= n; i++ {
		a[i] = i
	}

	a[1] = 0
	i := 2
	for i <= n {
		if a[i] != 0 {
			j := i + i
			for j <= n {
				a[j] = 0
				j = j + i
			}
		}
		i += 1
	}

	primes := make([]int, 0)
	for _, num := range a {
		if num != 0 {
			primes = append(primes, num)
		}
	}

	ism := []string{"Bobur", "Ali"}

	if n == 1 {
		fmt.Println("Ali")
	} else {
		fmt.Println(ism[len(primes)%2])
	}
}